camel/functions/search_functions.py

```diff
@@ -270,37 +333,69 @@ def summarize_text(text: str, query: str) -> str:
+def continue_search(query: str, answer: str) -> bool:
+    r"""Ask LLM whether to continue search or not.
 
     Args:
         query (str): Question you want to be answered.
+        answer (str): Answer to the query.
 
     Returns:
-        str: Summarized information from webs.
+        bool: True if the user want to continue search, False otherwise.
     """
-    # Google search will return a list of urls
-    responses = search_google(query)
+    prompt = TextPrompt("Do you think the ANSWER can answer the QUERY? "
+                        "Use only 'yes' or 'no' to answer\n"
+                        "===== QUERY ====={query}\n\n"
+                        "===== ANSWER ====={answer}")
+    prompt = prompt.format(query=query, answer=answer)
+    reply = prompt_single_step_agent(prompt)
+    if "yes" in str(reply).lower():
+        return False
+    return True
+
+
+def search_and_summarize(query: str, search_func: Callable) -> str:
+    """
+    Search the web for information using a specified search function
+    and summarize the results.
+
+    Args:
+        query (str): Question you want to be answered.
+        search_func (Callable): A function that takes a query.
+        and returns search results.
+
+    Returns:
+        str: Summarized information from the web.
+    """
+    responses = search_func(query)
     for item in responses:
         if "url" in item:
             url = item.get("url")
-            # Extract text
-            text = text_extract_from_web(str(url))
-            # Using chatgpt summarise text
-            answer = summarize_text(text, query)
-
-            # Let chatgpt decide whether to continue search or not
-            prompt = TextPrompt(
-                '''Do you think the answer: {answer} can answer the query:
-                {query}. Use only 'yes' or 'no' to answer.''')
-            prompt = prompt.format(answer=answer, query=query)
-            reply = prompt_single_step_agent(prompt)
-            if "yes" in str(reply).lower():
+            try:
+                extracted_text = text_extract_from_web(str(url))
+            except Exception:
+                continue
+            answer = summarize_text(extracted_text, query)
+            if not continue_search(query=query, answer=answer):
                 return answer
+    return "Failed to find an answer."
+
+
+def choose_search_and_summarize(query: str) -> str:
+    """
+    Choose between Google and DuckDuckGo search based on the availability
+    of Google API credentials, and summarize the results.
 
-    return "Failed to find the answer from google search."
+    Args:
+        query (str): Question you want to be answered.
+
+    Returns:
+        str: Summarized information from the web.
+    """
+    if os.getenv("GOOGLE_API_KEY") and os.getenv("SEARCH_ENGINE_ID"):
+        return search_and_summarize(query, search_google)
+    else:
+        return search_and_summarize(query, search_duckduckgo)
```

Comment:
why we need this function? actually we also have wolfram alpha in search function, if using function calling, I think agent would be able to switch the search function automaticlly